File: apps/api/app/models/organization.py
"""Organization model — top-level tenant entity.

Maps to shared-types: Organization, OrganizationSettings, OrganizationStatus,
SubscriptionPlan, IndustrySector, OrganizationSize.
"""

import enum
import uuid

from sqlalchemy import String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TimestampMixin, sa_enum


class OrganizationStatus(str, enum.Enum):
    ACTIVE = "active"
    SUSPENDED = "suspended"
    TRIAL = "trial"
    CHURNED = "churned"


class SubscriptionPlan(str, enum.Enum):
    FREE = "free"
    STARTER = "starter"
    PROFESSIONAL = "professional"
    ENTERPRISE = "enterprise"


class IndustrySector(str, enum.Enum):
    HEALTHCARE = "healthcare"
    RETAIL = "retail"
    MANUFACTURING = "manufacturing"
    SERVICES = "services"
    TECHNOLOGY = "technology"
    FINANCE = "finance"
    EDUCATION = "education"
    PUBLIC_SECTOR = "public_sector"
    HOSPITALITY = "hospitality"
    LOGISTICS = "logistics"
    OTHER = "other"


class OrganizationSize(str, enum.Enum):
    SMALL = "small"
    MEDIUM = "medium"
    LARGE = "large"
    ENTERPRISE = "enterprise"


class Organization(TimestampMixin, Base):
    """Organization — the root tenant entity.

    All other tenant-scoped entities reference organization.id
    via their organization_id column.
    """

    __tablename__ = "organizations"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    slug: Mapped[str] = mapped_column(
        String(100), unique=True, nullable=False, index=True
    )
    legal_name: Mapped[str | None] = mapped_column(String(255))
    siret: Mapped[str | None] = mapped_column(String(14))
    sector: Mapped[IndustrySector | None] = mapped_column(sa_enum(IndustrySector))
    size: Mapped[OrganizationSize | None] = mapped_column(sa_enum(OrganizationSize))
    headcount: Mapped[int | None] = mapped_column()
    status: Mapped[OrganizationStatus] = mapped_column(
        sa_enum(OrganizationStatus),
        default=OrganizationStatus.TRIAL,
    )
    plan: Mapped[SubscriptionPlan] = mapped_column(
        sa_enum(SubscriptionPlan),
        default=SubscriptionPlan.FREE,
    )
    timezone: Mapped[str] = mapped_column(String(50), default="Europe/Paris")
    locale: Mapped[str] = mapped_column(String(10), default="fr-FR")
    currency: Mapped[str] = mapped_column(String(3), default="EUR")
    contact_email: Mapped[str] = mapped_column(String(320), nullable=False)
    logo_url: Mapped[str | None] = mapped_column(Text)
    # JSONB for flexible settings (OrganizationSettings + WorkingDaysConfig)
    settings: Mapped[dict] = mapped_column(  # type: ignore[type-arg]
        JSONB, nullable=False, server_default="{}"
    )

    def __repr__(self) -> str:
        return f"<Organization {self.slug}>"

File: packages/shared-types/src/domain/organization.ts
// Organization domain types (multi-tenant)

import type { BaseEntity, UUID, ISODateString } from "../utils/common";
import type { WorkingDaysConfig } from "../utils/dates";

/** Organization status */
export type OrganizationStatus = "active" | "suspended" | "trial" | "churned";

/** Subscription plan */
export type SubscriptionPlan =
  | "free"
  | "starter"
  | "professional"
  | "enterprise";

/** Organization size category */
export type OrganizationSize = "small" | "medium" | "large" | "enterprise";

/** Industry sector */
export type IndustrySector =
  | "healthcare"
  | "retail"
  | "manufacturing"
  | "services"
  | "technology"
  | "finance"
  | "education"
  | "public_sector"
  | "hospitality"
  | "logistics"
  | "other";

/** Organization entity */
export interface Organization extends BaseEntity {
  /** Display name */
  name: string;
  /** URL-safe slug */
  slug: string;
  /** Legal entity name */
  legalName?: string;
  /** SIRET number (French companies) */
  siret?: string;
  /** Industry sector */
  sector?: IndustrySector;
  /** Organization size */
  size?: OrganizationSize;
  /** Total headcount */
  headcount?: number;
  /** Account status */
  status: OrganizationStatus;
  /** Subscription plan */
  plan: SubscriptionPlan;
  /** Trial end date */
  trialEndsAt?: ISODateString;
  /** Subscription start date */
  subscriptionStartedAt?: ISODateString;
  /** Working days configuration */
  workingDaysConfig: WorkingDaysConfig;
  /** Timezone (IANA) */
  timezone: string;
  /** Locale (BCP 47) */
  locale: string;
  /** Currency (ISO 4217) */
  currency: string;
  /** Primary contact email */
  contactEmail: string;
  /** Settings and preferences */
  settings: OrganizationSettings;
}

/** Organization settings */
export interface OrganizationSettings {
  /** Absence types enabled */
  absenceTypesEnabled: string[];
  /** Require manager approval */
  requireApproval: boolean;
  /** Approval workflow levels */
  approvalLevels: number;
  /** Enable forecasting */
  forecastingEnabled: boolean;
  /** Forecasting horizon (days) */
  forecastHorizonDays: number;
  /** Alert thresholds */
  alertThresholds: AlertThresholds;
  /** Data retention period (days) */
  dataRetentionDays: number;
  /** SSO configuration */
  ssoEnabled: boolean;
  /** SSO provider */
  ssoProvider?: "azure_ad" | "okta" | "google" | "custom_saml";
}

/** Alert threshold configuration */
export interface AlertThresholds {
  /** Understaffing risk threshold (%) */
  understaffingRisk: number;
  /** Absence rate threshold (%) */
  absenceRate: number;
  /** Consecutive absences threshold (days) */
  consecutiveAbsences: number;
  /** Forecast accuracy threshold (%) */
  forecastAccuracy: number;
}

/** Department/Team within an organization */
export interface Department extends BaseEntity {
  organizationId: UUID;
  name: string;
  code?: string;
  parentId?: UUID;
  managerId?: UUID;
  headcount: number;
  costCenter?: string;
  /** Minimum staffing level (%) */
  minStaffingLevel: number;
  /** Critical roles count */
  criticalRolesCount: number;
}

/** Site/Location */
export interface Site extends BaseEntity {
  organizationId: UUID;
  name: string;
  code?: string;
  address?: Address;
  timezone: string;
  /** Site-specific working days config */
  workingDaysConfig?: WorkingDaysConfig;
  headcount: number;
}

/** Address structure */
export interface Address {
  street: string;
  city: string;
  postalCode: string;
  country: string;
  region?: string;
}
